---
title: "Untitled"
author: "Quinones A"
date: "2024-04-15"
output: html_document
---
Chad recommended log transforming the RPKM and also stabilizing variance.

#setup
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(ggplot2); theme_set(theme_bw())
library(WGCNA)
library(gprofiler2)
library(ggpubr)
library(DESeq2)


morpho = read.csv("/Users/quino070/Bellwether_network/morpho_fpj.csv") %>%
  mutate(genotype=paste0("G", gen)) %>% dplyr::select(genotype, treatment,
                                                      replicate, 6:8)


ionomics = read.csv("/Users/quino070/Bellwether_network/ionomics_fpj.csv") %>%
  dplyr::select(genotype, treatment, replicate, 9:28) %>%
  mutate_at(4:ncol(.), log) %>% pivot_longer(4:23, names_to = "ion", values_to = "val") %>% 
  group_by(ion) %>% mutate(var=var(val)) %>% filter(var > 1) %>%
  pivot_wider(id_cols = 1:3, names_from = ion, values_from = val)

rnaseq = read.csv("raw_count_reads.csv")


```



#DESeq analysis
```{r}

mat = as.matrix(rnaseq[,-1])
rownames(mat) = rnaseq$gene_id

colData = as.data.frame(
  cbind(
    colnames(rnaseq)[-1],
    str_split(colnames(rnaseq)[-1], "_") %>% imap(~{.[2]}) %>% unlist())) %>% 
  column_to_rownames("V1")

colnames(colData) = "treatment"

dds = DESeqDataSetFromMatrix(mat, colData, design = ~treatment)

de_analysis = DESeq(dds)

res = results(de_analysis)

sig = subset(res,res$padj < 0.05)

dds_norm = vst
# Retrieve the normalized data from the `DESeqDataSet`
normalized_counts <- assay(dds_norm) %>%
  t() # Transpose this data


#write.csv(sig, "DE_genes.csv")



normalized_counts = normalized_counts %>% as.data.frame() %>% 
  rownames_to_column("id")  %>% rowwise() %>% mutate(
  genotype=str_split(id,"_") %>% unlist() %>% .[1],
  treatment=str_split(id,"_") %>% unlist() %>% .[2],
  replicate=str_split(id,"_") %>% unlist() %>% .[3]
) %>% dplyr::select(genotype, treatment, replicate, starts_with("Zm"))
```

```{r}
plotMA(res)
```

```{r}

```




```{do not run r message=FALSE, warning=FALSE}
# rnaseq = read.csv("/Users/quino070/Bellwether_network/normalized_reads_fpj.csv") %>%
#   group_by(gene_id) %>% mutate(var = var(RPKM), log_RPKM=log(RPKM) %>% round(3)) %>%
#   filter(var > 1) %>%
#   pivot_wider(id_cols = c("genotype", "treatment", "replicate"),
#               names_from = gene_id, values_from = RPKM)

```


```{r}
#df = rnaseq
df = normalized_counts

pc = df[,-c(1:3)] %>% prcomp() %>% .$x %>% as.data.frame() %>% mutate(treat = df$treatment, gen=df$genotype)

ggplot(pc, aes(PC1, PC2, color=treat, shape=gen)) + geom_point() 

```



```{r}
df = ionomics

pc = df[,-c(1:3)] %>% prcomp() %>% .$x %>% as.data.frame() %>% mutate(treat = df$treatment, gen=df$genotype)

ggplot(pc, aes(PC1, PC2, color=treat, shape=gen)) + geom_point() + xlim(c(-5,5)) +
  labs(color="Treatment", shape="Genotype", 
       x="Principal Component 1",
       y="Principal Component 2") + 
  scale_color_discrete(labels=c("Control", "Low Nitrogen")) + 
  scale_shape_discrete(labels=c("Ki3", "Mo18w", "Cml52", "Cml277")) +
  ggtitle("Principal component analysis for ionomics dataset after variance filtering")


ggsave("ionomics_pca.png", bg="transparent")



```

```{r}
normalized_counts = normalized_counts %>% 
  pivot_longer(cols=4:ncol(.), names_to = "gene_id", values_to = "norm_count") %>% 
  group_by(gene_id) %>% mutate(var=var(norm_count)) %>% filter(var < 1) %>% 
  dplyr::select(-var) %>% 
  pivot_wider(names_from = gene_id, values_from = norm_count)
```



# Merge rnaseq and ionomics dataframes together into one 
```{r}

master_df = left_join(normalized_counts %>% mutate(replicate=as.integer(replicate)), ionomics) 

#%>% right_join(morpho,.)
```


#Find a way to normalize all datasets 
```{r}
normalized_df = master_df %>% pivot_longer(cols=4:ncol(.), names_to = "var", 
                           values_to = "value") %>% 
  group_by(genotype, treatment, replicate) %>% 
  mutate(value = scale(value)) %>% pivot_wider(id_cols = 1:3, names_from = var, 
                                               values_from = value)
  
  
normalized_df
```

## Getting master dataset ready
```{r}
mat = as.matrix(normalized_df[,-c(1:3)])

rownames(mat) = normalized_df %>% mutate(id = paste0(genotype, "_",treatment, "_", replicate)) %>% .$id

## Determining parameters

#Input is the dataframe minus the first 3 columns

power = c(c(1:10), seq(from=12, to=50, by=2))
sft <- pickSoftThreshold(mat,
  powerVector = power,
  networkType = "signed",
  verbose=5)
```

```{r}
sft_df <- data.frame(sft$fitIndices) %>%
  dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)

ggplot(sft_df, aes(x = Power, y = model_fit, label = Power)) +
  # Plot the points
  geom_point() +
  # We'll put the Power labels slightly above the data points
  geom_text(nudge_y = 0.1) +
  # We will plot what WGCNA recommends as an R^2 cutoff
  geom_hline(yintercept = 0.80, col = "red") +
  # Just in case our values are low, we want to make sure we can still see the 0.80 level
  ylim(c(min(sft_df$model_fit), 1.05)) + xlim(c(0,30)) +
  # We can add more sensible labels for our axis
  xlab("Soft Threshold (power)") +
  ylab("Scale Free Topology Model Fit, signed R²") +
  ggtitle("Scale independence") +
  # This adds some nicer aesthetics to our plot
  theme_classic()

#ggsave("/Users/quino070/Bellwether_network/power_for_wgcna.svg", bg="transparent")
```


#running the WGCNA
```{r}
bwnet <- blockwiseModules(mat,
  maxBlockSize = 8000, # What size chunks (how many genes) the calculations should be run in
  TOMType = "signed", # topological overlap matrix
  power = 16, # soft threshold for network construction
  numericLabels = T, # Let's use numbers instead of colors for module labels
  mergeCutHeight=0.25,
  randomSeed = 1234, # there's some randomness associated with this calculation
  # so we should set a seed
)
```

```{r}
module_eigengenes <- bwnet$MEs

# Print out a preview
head(module_eigengenes)
```

Number of genes in each module
```{r}
table(bwnet$colors)
```


```{r}
plotDendroAndColors(bwnet$dendrograms[[1]], 
                    cbind(bwnet$unmergedColors, bwnet$colors), 
                    c("unmerged", "merged"),
                    dendrolabels=F, addGuide=T, hang=0.03, guideHang=0.05
                    )

```


Merge modules to morphological traits
```{r}
morpho_with_id = morpho %>% mutate(id = paste0(genotype, "_",treatment, "_", replicate)) %>% 
  dplyr::select(id, 4:6) %>% filter(id %in% rownames(module_eigengenes))

morpho_with_id_mat = morpho_with_id[,-1]
rownames(morpho_with_id_mat) = morpho_with_id$id

morpho_with_id_mat
```

Get correlation between morpho traits and modules
```{r}
cors = corAndPvalue(module_eigengenes, morpho_with_id_mat)
```

Heatmap of associations
```{r}
cor_df = merge(
  as.data.frame(cors$cor) %>% rownames_to_column("module") %>% pivot_longer(cols = 2:4, names_to = "morpho_var", values_to = "val"), 
  as.data.frame(cors$p) %>% rownames_to_column("module") %>% pivot_longer(cols = 2:4, names_to = "morpho_var", values_to = "pval")
) %>% mutate_at(3:4, ~{round(.x, 2)}) %>% 
  mutate(p = ifelse(pval<0.05, 
                    ifelse(pval < 0.01, "**", "*"), " "), val_p = paste(val, p)) 

cor_df %>% filter(module!="ME0") %>% 
  ggplot(aes(morpho_var, module, fill=val, label=val_p)) + 
  geom_tile() + geom_text() + scale_fill_gradient2(low = "red",
  mid = "white",
  high = "blue") + labs(
    x="Morphological trait", y="Module", fill="Correlation Coefficient"
  ) + scale_x_discrete(labels=c("Plant area (cm²)", 
                                "Plant height (cm)", "Mean plant hue (°)"),
                       expand = c(0,0)) +
  theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'), #transparent legend bg
    #legend.box.background = element_rect(fill='transparent') #transparent legend panel
  )

#ggsave("/Users/quino070/Bellwether_network/morpho_trait_cor_plot.png", bg = "transparent")
```


```{r}
blng = bwnet$colors %>% as.data.frame() %>% rename("module"=".") %>%
  rownames_to_column("variable")


full = master_df %>% 
  pivot_longer(cols=4:ncol(.), names_to = "variable", values_to = "value_in_sample") %>% 
  full_join(.,blng) %>% dplyr::select(1:4,6,5)

full
```

```{r}
cor_df %>% rename("correlation_coefficient"="val") %>% dplyr::select(1:4) 
  #write.csv(., "/Users/quino070/Bellwether_network/morpho_trait_correlations.csv", row.names = F)
```



1. What modules are significantly different between treatment and control samples?

2. GOanalysis: What are the genes in each mod







#GO analysis

```{r}
get_version_info(organism = "zmays")

full = read.csv("master_df_before_normalization.csv") 

m23_genes = full %>% filter(module==23) %>% .$variable

gostres = gost(query=m23_genes, organism = "zmays")

gostres


gostplot(gostres, capped=F, interactive=F)
```

```{r}
m23_genes = full %>% filter(module==23) %>% .$variable

gostres = gost(query=m23_genes, organism = "zmays",correction_method = "fdr")

gostres


gostplot(gostres, capped=F, interactive=F)


mod_vec = full$module %>% unique()

enrichment = imap(mod_vec, ~{
  no_ions = full %>% filter(module==.x) %>% .$variable 
  gost(query=no_ions, organism = "zmays", correction_method = "fdr") %>% .[["result"]] %>% cbind(module=.x)
})

GO_terms_for_modules = enrichment %>% do.call("rbind",.) %>% 
  rowwise() %>% mutate(parents = unlist(parents) %>% paste(., collapse=","))



#write.csv(GO_terms_for_modules, "GO_enrichment.csv", row.names = F)

```


```{r}
plots = imap(0:24, ~{
  GO_terms_for_modules %>% mutate(GeneRatio = intersection_size/query_size) %>% 
  filter(p_value < 0.01 & module==.x) %>% 
  ggplot(aes(log(p_value), term_name, size=GeneRatio, color=source)) +
  geom_point() + labs(y="Term", fill="GO term source", x="log(adjusted p value)") + 
  scale_fill_discrete(labels=c("Biological Process", "Cellular Compartment"))
})

```







#Getting significantly different modules between C and N
```{r}
bwnet = readRDS("wgcna_fpj.RDS")

module_eigengenes <- bwnet$MEs

module_eigengenes = 
module_eigengenes %>% rownames_to_column("id") %>%
  #pivot_longer(2:ncol(.), names_to = "module", values_to = "eigengene") %>%
  rowwise() %>% mutate(
    genotype=str_split(id, "_") %>% unlist() %>% .[1],
    treatment=str_split(id, "_") %>% unlist() %>% .[2]
  ) %>% dplyr::select(27,28, 2:26) %>% mutate(treatment = as.factor(treatment))

mods = colnames(module_eigengenes) %>% .[-c(1:2)]
```



```{r}

CvsN = imap(mods, ~{
  sum = lm(eval(parse(text=.x)) ~ treatment, module_eigengenes) %>% summary()
  sum[["coefficients"]] %>% as.data.frame() %>% cbind(module=.x)}) %>% 
  do.call("rbind",.) %>% 
  mutate(significance = ifelse(`Pr(>|t|)`<0.05, 
                    ifelse(`Pr(>|t|)` < 0.01, "**", "*"), " ")) 

#write.csv(CvsN, "module_significance_for_treatment_fpj.csv", row.names = F)


module_eigengenes %>% 
  pivot_longer(3:ncol(.), names_to = "Module", values_to = "eigengenes") %>% 
  filter(Module %in% 
           c("ME6", "ME10", "ME13", "ME19", "ME12", 
             "ME15","ME17", "ME1", "ME0")) %>% 
  ggplot(aes(Module,eigengenes, fill=treatment)) + geom_boxplot() +
  stat_compare_means(method="t.test", aes(label=..p.signif..)) +
  labs(y="Eigengenes", fill="Treatment") + scale_fill_discrete(labels=c("Control", "Low Nitrogen")) + 
  ylim(c(-0.4, 0.6)) + theme_classic() +
  theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'), #transparent legend bg
    #legend.box.background = element_rect(fill='transparent') #transparent legend panel
  )


#ggsave("significantly_different_modules.png",bg = "transparent")
```



limma for ionomics


rnaseq-> 
log transofmr the RPKM
filter by variance


normalize everything by length 



```{r}
GO_terms_for_modules = read.csv("GO_enrichment.csv")
GO_terms_for_modules
```

```{r}
goplots  = imap(0:24, ~{
  GO_terms_for_modules %>% mutate(GeneRatio = intersection_size/query_size) %>%
    filter(module==.x) %>%
    ggplot(aes(log(p_value), term_name, size=GeneRatio, color=source)) +
    geom_point() + labs(y="Term", fill="GO term source", x="log(adjusted p value)") +
    scale_fill_discrete(labels=c("Biological Process", "Cellular Compartment")) +
    ggtitle(paste0("Module ", .x))
})

```
down-regulated = 13,15,17,19
up-regulated = 0,10,12,6


```{r}
GO_terms_for_modules %>% 
  mutate(GeneRatio = intersection_size/query_size, 
         regulation = ifelse(module %in% c(13,15,17,19), "down-regulated", 
                             ifelse(module %in% c(0,10,12,6), "up-regulated", "not significant")
                             )
         ) %>% filter(source=="GO:BP" & regulation != "not significant") %>% 
  group_by(regulation, term_name) %>% 
  summarise(GeneRatio=mean(GeneRatio), p_value=mean(p_value)) %>% 
  top_n(5, wt = GeneRatio) %>% arrange(regulation) %>% 
  ggplot(aes(regulation, term_name, size=GeneRatio, color=log(p_value))) +
  geom_point() + 
  scale_color_viridis_c(direction = -1) + ggtitle("Top 5 GO terms")

#ggsave("top_5_GO_terms.svg", bg="transparent")
```

```{r}
DE_genes = read.csv("DE_genes.csv") %>% rename("X"="gene_id")


DE_genes
```

```{r}
DE = left_join(DE_genes, full %>% rename("variable"="gene_id") %>% 
                 dplyr::select(4:5) %>% unique()) %>% 
  mutate(regulation = ifelse(module %in% c(13,15,17,19), "Down-regulated", 
                             ifelse(module %in% c(0,10,12,6), 
                                    "Up-regulated", "Not significant")))




```
down-regulated = 13,15,17,19
up-regulated = 0,10,12,6

```{r}
DE %>% filter(log2FoldChange > 5) %>% 
  ggplot(aes(log2FoldChange, gene_id, fill=regulation, group=module)) + 
  geom_bar(stat="identity") +
  geom_bar(data=DE %>% filter(log2FoldChange < -5), stat="identity") + 
  scale_fill_manual(values=c("#421049","#BBB8B8", "#194419")) +
  theme_classic() + 
  labs(fill="Module regulation", x="Fold change (log2 scale)", y="Gene") +
  #theme(axis.text.y = element_text(size=3)) + 
  ggtitle("DE genes with over 5 log2(Fold Change) and their module significance")+
      theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'), #transparent legend bg
    #legend.box.background = element_rect(fill='transparent') #transparent legend panel
  )
#ggsave("DE_genes_fc5_plus_module_significance.png", bg="transparent")
```


```{r}

downreg = DE %>% top_n(10, wt = -log2FoldChange)
down = gost(query=downreg$gene_id, organism = "zmays", correction_method = "fdr") 
down[["result"]] %>% arrange(-intersection_size)

```
Highly downregulated genes involved mainly in oxidoreductase activity, protein S-S reductase activity, S-S oxidoreductase activity, ELECTRON TRANSFER, a bunch of oxidoreductase. No BP terms enriched



```{r}
upreg=DE %>% top_n(10, wt = log2FoldChange)
up = gost(query=upreg$gene_id, organism = "zmays", significant = F) 
up[["result"]] %>% arrange(-intersection_size)
```

No significant terms enriched. Non significant terms enriched: 



```{r}
summary(res)
```


Most down-regulated gene:
	
Zm00001eb132640: 
 	organic cyclic compound biosynthetic process 
 	sterol metabolic process
  In arabidopsis: Encodes a member of the SMO1 family of sterol 4alpha-methyl oxidases. More specifically functions as a 4,4-dimethyl-9beta,19-cyclopropylsterol-4alpha-methyl oxidase. Works together with SMO1-2 to maintain correct sterol composition and balance auxin and cytokinin activities during embryogenesis.
  So this one is an enzyme
  NEW GENE IN THE V5 version of the genome
  

Most up-regulated gene in a module:
Zm00001eb423830
  pteridine-containing compound metabolic process 
  integral component of membrane
  this one is a membrane protein
  WITH SNPs TRAITS ASSOCIATED: https://www.maizegdb.org/data_center/reference/9025599 Wallace, JG, et al. 2014. PLoS Genetics. 10:e1004845
  KERNEL ROW NUMBER AND PHOTOPERIOD GROWING DEGREE DAYS TO ANTHESIS


```{r}
go = read.csv("GO_enrichment.csv")

go_plot = go %>% 
  mutate(GeneRatio = intersection_size/query_size, 
         regulation = ifelse(module %in% c(13,15,17,19), "Down-regulated", 
                             ifelse(module %in% c(0,10,12,6), "Up-regulated", "not significant")
                             )
         ) %>% filter(module %in% c(10,12,13,15,17,19,6) & GeneRatio > 0.2) %>% 
  group_by(module) %>% slice_max(n=5, order_by=GeneRatio) %>% 
  filter(!term_name %in% c("small molecule metabolic process", 
                           "intracellular anatomical structure"))



 go_plot %>%  ggplot(aes(regulation, fct_reorder(term_name, -module), 
             size=GeneRatio, color=log(p_value))) + 
  geom_point() + scale_color_viridis_c(direction = -1) + 
  labs(y="GO term", x="Regulation", color="log(p.adj)", size="Gene Ratio") +
    theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'), #transparent legend bg
    #legend.box.background = element_rect(fill='transparent') #transparent legend panel
  ) + guides(size=guide_legend(order=1))


#ggsave("top5_enriched_go_terms_per_module.png", width = 7,bg="transparent")

```





